README.rst
webthing
========


.. image:: https://img.shields.io/badge/license-MPL--2.0-blue.svg
    :target: https://github.com/WebThingsIO/webthing-python/blob/master/LICENSE.txt

Implementation of an HTTP `Web Thing <https://iot.mozilla.org/wot/>`_. This library is compatible with 3.5+.

m2ag-labs modifications:
    Adds HS256 JWT Authorization to webthings-python.
    By default JWT authentication is enabled.
    Config file in .m2ag-labs/secrets. Server will generate a default configuration at first run.
    Script to generate JWT in examples folder (run server first).
    See example/multiple-things-jet.py for an example.


Installation
============

``webthing`` must be installed manually.

First install the dependencies:

.. code:: shell

  $ pip install pyjwt ifaddr jsonschema pyee tornado zeroconf

Then clone the repository as your project directory (only webthings directory is needed):

.. code:: shell

  $ git clone https://github.com/m2ag-labs/webthing-python.git


JWT Control
===========

JWT use is configured via ~/.webthings/things/jwt_secret.json

.. code:: json

    {
      "enable": true,
      "secret_key": "some real long string - qwesovk4qzbp5o3vudxp7au131mjkncgqrx93xs18im2dexqhpues1fxz5ju3grdhn7nqhm4hjl0ns4e",
      "auth_header": "Authorization",
      "auth_param": "jwt",
      "auth_error_code": 401,
      "auth_error_thing": {
        "id": "urn:<your org>:security:authorization-required",
        "title": "<add instructive title - this shows up in the gateway>",
        "@context": "https://webthings.io/schemas",
        "description": "Bearer tokens are required for this device",
        "securityDefinitions": {
          "bearer_sc": {
            "scheme": "bearer",
            "alg": "HS256",
            "description": "Security is required for this thing.",
            "authorization": "https://holly.local:8443/auth.html"
          }
        },
        "security": [
          "bearer_sc"
        ]
      },
      "jwt_options": {
        "verify_signature": true,
        "verify_exp": false,
        "verify_nbf": false,
        "verify_iat": true,
        "verify_aud": false
      }
    }

First install the dependencies:

.. code:: shell

  $ pip install pyjwt ifaddr jsonschema pyee tornado zeroconf

Then clone the repository as your project directory (only webthings directory is needed):
.. code:: shell

  $ git clone https://github.com/m2ag-labs/webthing-python.git


Running the Sample
==================

Change to your project directory and run:

.. code:: shell

   $ python3 example/single-thing.py

This starts a server and lets you search for it from your gateway through mDNS. To add it to your gateway, navigate to the Things page in the gateway's UI and click the + icon at the bottom right. If both are on the same network, the example thing will automatically appear.

Example Implementation
======================

In this code-walkthrough we will set up a dimmable light and a humidity sensor (both using fake data, of course). Both working examples can be found in the `examples directory <https://github.com/WebThingsIO/webthing-python/tree/master/example>`_.

Dimmable Light
--------------

Imagine you have a dimmable light that you want to expose via the web of things API. The light can be turned on/off and the brightness can be set from 0% to 100%. Besides the name, description, and type, a |Light|_ is required to expose two properties:

.. |Light| replace:: ``Light``
.. _Light: https://iot.mozilla.org/schemas/#Light

* ``on``: the state of the light, whether it is turned on or off

  - Setting this property via a ``PUT {"on": true/false}`` call to the REST API toggles the light.

* ``brightness``: the brightness level of the light from 0-100%

  - Setting this property via a PUT call to the REST API sets the brightness level of this light.

First we create a new Thing:

.. code:: python

    light = Thing(
        'urn:dev:ops:my-lamp-1234',
        'My Lamp',
        ['OnOffSwitch', 'Light'],
        'A web connected lamp'
    )

Now we can add the required properties.

The ``on`` property reports and sets the on/off state of the light. For this, we need to have a ``Value`` object which holds the actual state and also a method to turn the light on/off. For our purposes, we just want to log the new state if the light is switched on/off.

.. code:: python

  light.add_property(
      Property(
          light,
          'on',
          Value(True, lambda v: print('On-State is now', v)),
          metadata={
              '@type': 'OnOffProperty',
              'title': 'On/Off',
              'type': 'boolean',
              'description': 'Whether the lamp is turned on',
          }))

The ``brightness`` property reports the brightness level of the light and sets the level. Like before, instead of actually setting the level of a light, we just log the level.

.. code:: python

  light.add_property(
      Property(
          light,
          'brightness',
          Value(50, lambda v: print('Brightness is now', v)),
          metadata={
              '@type': 'BrightnessProperty',
              'title': 'Brightness',
              'type': 'number',
              'description': 'The level of light from 0-100',
              'minimum': 0,
              'maximum': 100,
              'unit': 'percent',
          }))

Now we can add our newly created thing to the server and start it:

.. code:: python

  # If adding more than one thing, use MultipleThings() with a name.
  # In the single thing case, the thing's name will be broadcast.
  server = WebThingServer(SingleThing(light), port=8888)

  try:
      server.start()
  except KeyboardInterrupt:
      server.stop()

This will start the server, making the light available via the WoT REST API and announcing it as a discoverable resource on your local network via mDNS.

Sensor
------

Let's now also connect a humidity sensor to the server we set up for our light.

A |MultiLevelSensor|_ (a sensor that returns a level instead of just on/off) has one required property (besides the name, type, and optional description): ``level``. We want to monitor this property and get notified if the value changes.

.. |MultiLevelSensor| replace:: ``MultiLevelSensor``
.. _MultiLevelSensor: https://iot.mozilla.org/schemas/#MultiLevelSensor

First we create a new Thing:

.. code:: python

  sensor = Thing(
      'urn:dev:ops:my-humidity-sensor-1234',
      'My Humidity Sensor',
       ['MultiLevelSensor'],
       'A web connected humidity sensor'
  )

Then we create and add the appropriate property:

* ``level``: tells us what the sensor is actually reading

  - Contrary to the light, the value cannot be set via an API call, as it wouldn't make much sense, to SET what a sensor is reading. Therefore, we are creating a **readOnly** property.

    .. code:: python

      level = Value(0.0);

      sensor.add_property(
          Property(
              sensor,
              'level',
              level,
              metadata={
                  '@type': 'LevelProperty',
                  'title': 'Humidity',
                  'type': 'number',
                  'description': 'The current humidity in %',
                  'minimum': 0,
                  'maximum': 100,
                  'unit': 'percent',
                  'readOnly': True,
              }))

Now we have a sensor that constantly reports 0%. To make it usable, we need a thread or some kind of input when the sensor has a new reading available. For this purpose we start a thread that queries the physical sensor every few seconds. For our purposes, it just calls a fake method.

.. code:: python

  self.sensor_update_task = \
      get_event_loop().create_task(self.update_level())

  async def update_level(self):
      try:
          while True:
              await sleep(3)
              new_level = self.read_from_gpio()
              logging.debug('setting new humidity level: %s', new_level)
              self.level.notify_of_external_update(new_level)
      except CancelledError:
          pass

This will update our ``Value`` object with the sensor readings via the ``self.level.notify_of_external_update(read_from_gpio())`` call. The ``Value`` object now notifies the property and the thing that the value has changed, which in turn notifies all websocket listeners.

Adding to Gateway
=================

To add your web thing to the WebThings Gateway, install the "Web Thing" add-on and follow the instructions `here <https://github.com/WebThingsIO/thing-url-adapter#readme>`_.


CHANGELOG.md
# webthing Changelog

## [Unreleased]

## [0.15.1] - 2021-03-13
### Added
- Forked at 15.0.0
- Parameter to disable host validation in server moved to options object.
- Add JWT support and examples

## [0.15.0] - 2021-01-02
### Added
- Parameter to disable host validation in server.

## [0.14.0] - 2020-09-23
### Changed
- Dropped Python 2.7 support.
- Update author and URLs to indicate new project home.

## [0.13.2] - 2020-06-18
### Changed
- mDNS record now indicates TLS support.

## [0.13.1] - 2020-05-29
### Changed
- Support new zeroconf version

## [0.13.0] - 2020-05-04
### Changed
- Invalid POST requests to action resources now generate an error status.
- Python 3.4 support dropped

## [0.12.2] - 2020-03-27
### Added
- Support OPTIONS requests to allow for CORS.

## [0.12.1] - 2020-01-13
### Changed
- Abstracted WebSocket subscriber class to allow library to work with other web frameworks.

## [0.12.0] - 2019-07-12
### Changed
- Things now use `title` rather than `name`.
- Things now require a unique ID in the form of a URI.
### Added
- Ability to set a base URL path on server.
- Support for `id`, `base`, `security`, and `securityDefinitions` keys in thing description.

## [0.11.3] - 2019-04-10
### Changed
- Simpler dependencies with no native requirements.

## [0.11.2] - 2019-03-11
### Added
- Support for Tornado 6.x

## [0.11.1] - 2019-01-28
### Added
- Support for Python 2.7 and 3.4

## [0.11.0] - 2019-01-16
### Changed
- WebThingServer constructor can now take a list of additional API routes.
### Fixed
- Properties could not include a custom `links` array at initialization.

## [0.10.0] - 2018-11-30
### Changed
- Property, Action, and Event description now use `links` rather than `href`. - [Spec PR](https://github.com/WebThingsIO/wot/pull/119)

[Unreleased]: https://github.com/WebThingsIO/webthing-python/compare/v0.15.0...HEAD
[0.15.0]: https://github.com/WebThingsIO/webthing-python/compare/v0.14.0...v0.15.0
[0.14.0]: https://github.com/WebThingsIO/webthing-python/compare/v0.13.2...v0.14.0
[0.13.2]: https://github.com/WebThingsIO/webthing-python/compare/v0.13.1...v0.13.2
[0.13.1]: https://github.com/WebThingsIO/webthing-python/compare/v0.13.0...v0.13.1
[0.13.0]: https://github.com/WebThingsIO/webthing-python/compare/v0.12.2...v0.13.0
[0.12.2]: https://github.com/WebThingsIO/webthing-python/compare/v0.12.1...v0.12.2
[0.12.1]: https://github.com/WebThingsIO/webthing-python/compare/v0.12.0...v0.12.1
[0.12.0]: https://github.com/WebThingsIO/webthing-python/compare/v0.11.3...v0.12.0
[0.11.3]: https://github.com/WebThingsIO/webthing-python/compare/v0.11.2...v0.11.3
[0.11.2]: https://github.com/WebThingsIO/webthing-python/compare/v0.11.1...v0.11.2
[0.11.1]: https://github.com/WebThingsIO/webthing-python/compare/v0.11.0...v0.11.1
[0.11.0]: https://github.com/WebThingsIO/webthing-python/compare/v0.10.0...v0.11.0
[0.10.0]: https://github.com/WebThingsIO/webthing-python/compare/v0.9.2...v0.10.0


webthing/jwtauth.py
"""
    JSON Web Token auth for Tornado
    Modified by marc at m2ag.labs (marc@m2ag.net) from files found here:
    https://github.com/paulorodriguesxv/tornado-json-web-token-jwt
    Added configuration from file, errors as dict, added a check for
    websocket upgrade (wss add auth parameter to connect string --
    ?jwt=<token>)
"""
import jwt
import json
import socket
from pathlib import Path

"""
    Secret can be any string -- will be auto generated if jwt_secrets file does not exist
    Use the secret to generate the JWT.
    If it does not exist .m2ag-labs/secrets/jwt_config.json will be created.
    Setting enable false will signal the thing to bypass auth checks.
"""
CONFIG_FILE = 'jwt_config.json'
# CONFIG_PATH = f'{str(Path.home())}/.m2ag-labs/secrets'
CONFIG_PATH = f'{str(Path.home())}/.webthings/things'
OPTIONS = {
    'enable': True,
    'local_bypass': True,
    'secret_key': '',
    'auth_header': 'Authorization',
    'auth_param': 'jwt',
    'auth_error_code': 401,
    'auth_error_thing': {  # this needs to be json
        "id": "urn:m2ag:security:authorization-required",
        "title": f"{socket.gethostname()} is a secure thing. See https://{socket.gethostname()}.local:8443/auth.html",
        "@context": "https://webthings.io/schemas",
        "description": "Bearer tokens are required for this device",
        "securityDefinitions": {
            "bearer_sc": {
                "scheme": "bearer",
                "alg": "HS256",
                "description": "Security is required for this thing.",
                "authorization": f"https://{socket.gethostname()}.local:8443/auth.html"
            }
        },
        "security": ["bearer_sc"]
    },
    'jwt_options': {
        'verify_signature': True,
        'verify_exp': False,  # JWTs will never expire for this device if False
        'verify_nbf': False,
        'verify_iat': True,
        'verify_aud': False
    }
}


def get_options():
    try:
        with open(f'{CONFIG_PATH}/{CONFIG_FILE}', 'r') as file:
            opts = json.loads(file.read().replace('\n', ''))
            for i in opts:
                OPTIONS[i] = opts[i]
            del opts  # clean up

    except FileNotFoundError:
        import string
        import random

        rando = string.ascii_lowercase + string.digits
        OPTIONS['secret_key'] = ''.join(random.choice(rando) for i in range(100))

        Path(CONFIG_PATH).mkdir(parents=True, exist_ok=True)
        with open(f'{CONFIG_PATH}/{CONFIG_FILE}', 'w') as file:
            file.write(json.dumps(OPTIONS))


get_options()


def return_auth_error(handler, message):
    """
        Return authorization error
    """
    handler._transforms = []
    handler.set_status(OPTIONS['auth_error_code'])
    handler.write(OPTIONS['auth_error_thing'])
    handler.finish()


def return_header_error(handler):
    return_auth_error(handler, OPTIONS['auth_error_thing'])


def jwtauth(handler_class):
    # Tornado JWT Auth Decorator

    def wrap_execute(handler_execute):
        def require_auth(handler):
            # configure the jwt with a config file
            if not OPTIONS['enable']:
                return True
            # TODO: do we need to allow this?
            # makes same host access harder
            if OPTIONS['local_bypass'] and handler.request.remote_ip == '127.0.0.1':
                return True
            auth = handler.request.headers.get(OPTIONS['auth_header'])
            if auth:
                parts = auth.split()
                try:
                    jwt.decode(
                        parts[1],
                        OPTIONS['secret_key'],
                        algorithms=["HS256"],
                        options=OPTIONS['jwt_options']
                    )
                except Exception as err:
                    print(str(err))
                    return_auth_error(handler, OPTIONS['auth_error_thing'])

            else:
                # is this websocket upgrade? if so look for auth header in
                # params
                upgrade = handler.request.headers.get("Upgrade")
                if upgrade == 'websocket':
                    # broken up for length issue (flake8)
                    handle = handler.request.query_arguments
                    auth = handle.get(OPTIONS['auth_param'])
                    if auth:
                        try:
                            jwt.decode(
                                auth[0].decode('UTF-8'),
                                OPTIONS['secret_key'],
                                algorithms=["HS256"],
                                options=OPTIONS['jwt_options']
                            )
                        except Exception as err:
                            print(str(err))
                            return_auth_error(handler, OPTIONS['auth_error_thing'])
                        return True

                handler._transforms = []
                handler.write(OPTIONS['auth_error_thing'])
                handler.finish()

            return True

        def _execute(self, transforms, *args, **kwargs):

            try:
                require_auth(self)
            except Exception:
                return False

            return handler_execute(self, transforms, *args, **kwargs)

        return _execute

    handler_class._execute = wrap_execute(handler_class._execute)
    return handler_class


example/generate_jwt.py
from datetime import datetime, timedelta
from pathlib import Path

from json import loads
from jwt import encode

# secret key in config
# if file read fails, create a file with a random string in it. (thing uses same file)
# secret can be any string
# enable false will bypass auth checks'''
try:
    #  with open(f'{str(Path.home())}/.m2ag-labs/secrets/jwt_secret.json', 'r') as file:
    with open(f'{str(Path.home())}/.webthings/things/jwt_secret.json', 'r') as file:
        opts = loads(file.read().replace('\n', ''))
        for i in opts:
            if i == 'secret':
                SECRET = opts[i]
except FileNotFoundError:
    print('Please run the sever with JWT enabled before using this script')
    quit()

AUTH_TTL = 31104000  # this is super long for development

"""
    Encode a new token with JSON Web Token (PyJWT)
"""
encoded = encode({
    'context': 'reserved for future use',
    'exp': datetime.utcnow() + timedelta(seconds=AUTH_TTL)},
    SECRET,
    algorithm='HS256'
)
# raspian desktop full -- this returns a buffer and needs to be decoded - currently pyjwt 1.7.0
# raspian lite -- returns a string -- version pyjwt 2.0.0
if isinstance(encoded, str):
    print(encoded)
else:
    print(encoded.decode('ascii'))


example/multiple-things-jwt.py
from __future__ import division, print_function
from webthing import (Action, Event, MultipleThings, Property, Thing, Value,
                      WebThingServer)
import logging
import random
import time
import tornado.ioloop
import uuid


class OverheatedEvent(Event):

    def __init__(self, thing, data):
        Event.__init__(self, thing, 'overheated', data=data)


class FadeAction(Action):

    def __init__(self, thing, input_):
        Action.__init__(self, uuid.uuid4().hex, thing, 'fade', input_=input_)

    def perform_action(self):
        time.sleep(self.input['duration'] / 1000)
        self.thing.set_property('brightness', self.input['brightness'])
        self.thing.add_event(OverheatedEvent(self.thing, 102))


class ExampleDimmableLight(Thing):
    """A dimmable light that logs received commands to stdout."""

    def __init__(self):
        Thing.__init__(
            self,
            'urn:dev:ops:my-lamp-1234',
            'My Lamp',
            ['OnOffSwitch', 'Light'],
            'A web connected lamp'
        )

        self.add_property(
            Property(self,
                     'on',
                     Value(True, lambda v: print('On-State is now', v)),
                     metadata={
                         '@type': 'OnOffProperty',
                         'title': 'On/Off',
                         'type': 'boolean',
                         'description': 'Whether the lamp is turned on',
                     }))

        self.add_property(
            Property(self,
                     'brightness',
                     Value(50, lambda v: print('Brightness is now', v)),
                     metadata={
                         '@type': 'BrightnessProperty',
                         'title': 'Brightness',
                         'type': 'integer',
                         'description': 'The level of light from 0-100',
                         'minimum': 0,
                         'maximum': 100,
                         'unit': 'percent',
                     }))

        self.add_available_action(
            'fade',
            {
                'title': 'Fade',
                'description': 'Fade the lamp to a given level',
                'input': {
                    'type': 'object',
                    'required': [
                        'brightness',
                        'duration',
                    ],
                    'properties': {
                        'brightness': {
                            'type': 'integer',
                            'minimum': 0,
                            'maximum': 100,
                            'unit': 'percent',
                        },
                        'duration': {
                            'type': 'integer',
                            'minimum': 1,
                            'unit': 'milliseconds',
                        },
                    },
                },
            },
            FadeAction)

        self.add_available_event(
            'overheated',
            {
                'description':
                    'The lamp has exceeded its safe operating temperature',
                'type': 'number',
                'unit': 'degree celsius',
            })


class FakeGpioHumiditySensor(Thing):
    """A humidity sensor which updates its measurement every few seconds."""

    def __init__(self):
        Thing.__init__(
            self,
            'urn:dev:ops:my-humidity-sensor-1234',
            'My Humidity Sensor',
            ['MultiLevelSensor'],
            'A web connected humidity sensor'
        )

        self.level = Value(0.0)
        self.add_property(
            Property(self,
                     'level',
                     self.level,
                     metadata={
                         '@type': 'LevelProperty',
                         'title': 'Humidity',
                         'type': 'number',
                         'description': 'The current humidity in %',
                         'minimum': 0,
                         'maximum': 100,
                         'unit': 'percent',
                         'readOnly': True,
                     }))

        logging.debug('starting the sensor update looping task')
        self.timer = tornado.ioloop.PeriodicCallback(
            self.update_level,
            3000
        )
        self.timer.start()

    def update_level(self):
        new_level = self.read_from_gpio()
        logging.debug('setting new humidity level: %s', new_level)
        self.level.notify_of_external_update(new_level)

    def cancel_update_level_task(self):
        self.timer.stop()

    @staticmethod
    def read_from_gpio():
        """Mimic an actual sensor updating its reading every couple seconds."""
        return abs(70.0 * random.random() * (-0.5 + random.random()))


def run_server():
    # Create a thing that represents a dimmable light
    light = ExampleDimmableLight()

    # Create a thing that represents a humidity sensor
    sensor = FakeGpioHumiditySensor()

    # If adding more than one thing, use MultipleThings() with a name.
    # In the single thing case, the thing's name will be broadcast.
    # WS auth is a little different and described by bearer_ws
    # Add "disable_host_validation" : True here.
    options = {
        'security': 'bearer_sc',
        'security_ws': 'bearer_ws',
        'security_definitions': {
            "bearer_sc": {
                "scheme": "bearer",
                "alg": "HS256",
                "description": "A token must be configured for this thing",
                "authorization": "https://holly.local:8443/auth.html"
            },
            "bearer_ws": {
                "scheme": "bearer",
                "alg": "HS256",
                "in": "query",
                "name": "jwt",
                "description": "A token must be configured for this thing",
                "authorization": "https://holly.local:8443/auth.html"
            }
        }
    }

    server = WebThingServer(MultipleThings([light, sensor],
                                           'LightAndTempDevice'),
                            port=8888, options=options)
    try:
        logging.info('starting the server')
        server.start()
    except KeyboardInterrupt:
        logging.debug('canceling the sensor update looping task')
        sensor.cancel_update_level_task()
        logging.info('stopping the server')
        server.stop()
        logging.info('done')


if __name__ == '__main__':
    logging.basicConfig(
        level=10,
        format="%(asctime)s %(filename)s:%(lineno)s %(levelname)s %(message)s"
    )
    run_server()
